File: Programmers/DFSorBFS/사라지는_발판.js
let n, m;
let dir = [ [0, 1], [0, -1], [-1, 0], [1, 0] ];

function isInside(x, y) {
    return 0 <= x && x < n && 0 <= y && y < m;
}

function canMove(board, x, y) {
    for(let i=0; i<4; i++) {
        let nx = x + dir[i][0];
        let ny = y + dir[i][1];
        if(isInside(nx, ny) && board[nx][ny] === 1) {
            return true;
        }
    }
    return false;
}

function recur(board, ax, ay, bx, by) {
    if(!canMove(board, ax, ay)) {
        return [false, 0];
    }
    
    const result = [false, 0];
    if(board[ax][ay] === 1) {
        let max = 0;
        let min = 987654321;
        
        for(let i=0; i<4; i++) {
            let nx = ax + dir[i][0];
            let ny = ay + dir[i][1];
            
            if(isInside(nx, ny) && board[nx][ny] === 1) {
                board[ax][ay] = 0;
                const [win, dist] = recur(board, bx, by, nx, ny);
                board[ax][ay] = 1;
                
                if(!win) {
                    result[0] = true;
                    min = Math.min(min, dist);
                } else if(!result[0]) {
                    max = Math.max(max, dist);
                }
            }   
        }
        result[1] = result[0] ? min + 1 : max + 1;
    }
    return result;
}

function solution(board, aloc, bloc) {
    n = board.length;
    m = board[0].length;
    return recur(board, aloc[0], aloc[1], bloc[0], bloc[1])[1];
}

File: Programmers/DFSorBFS/사라지는_발판.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <utility>

using namespace std;

int N, M, answer;
int dx[4] = {0,0,-1,1};
int dy[4] = {1,-1,0,0};

bool is_inside(int x, int y) {
    return 0 <= x && x < N && 0 <= y && y < M;
}

bool can_move(vector<vector<int>> board, int x, int y) {
    for(int dir=0; dir<4; dir++) {
        int nx = x + dx[dir];
        int ny = y + dy[dir];
        if(is_inside(nx,ny) && board[nx][ny] == 1) {
            return true;
        }
    }
    return false;
}

pair<bool, int> move(vector<vector<int>> board, int ax, int ay, int bx, int by) {
    if(board[ax][ay] == 0 || !can_move(board, ax, ay)) {
        return {false, 0};
    }
    
    int MIN = 987654321;
    int MAX = 0;
    pair<bool, int> result = {false, 0};
    
    for(int dir=0; dir<4; dir++) {
        int nx = ax + dx[dir];
        int ny = ay + dy[dir];
        if(is_inside(nx,ny) && board[nx][ny] == 1) {
            board[ax][ay] = 0;
            pair<bool, int> next = move(board, bx, by, nx, ny);
            board[ax][ay] = 1;
            
            if(!next.first) {
                result.first = true;
                MIN = min(MIN, next.second);
            } else if(!result.first) {
                MAX = max(MAX, next.second);
            }
        }
    }
    result.second = result.first ? MIN + 1 : MAX + 1;
    return result;
}

int solution(vector<vector<int>> board, vector<int> aloc, vector<int> bloc) {
    N = (int) board.size();
    M = (int) board[0].size();
    
    return move(board, aloc[0], aloc[1], bloc[0], bloc[1]).second;
}
